libraries/shared/watcher/storage_watcher.go

```diff
@@ -18,62 +18,92 @@ package watcher
 
 import (
 	"fmt"
+	"time"
+
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/sirupsen/logrus"
+
 	"github.com/vulcanize/vulcanizedb/libraries/shared/fetcher"
 	"github.com/vulcanize/vulcanizedb/libraries/shared/storage"
 	"github.com/vulcanize/vulcanizedb/libraries/shared/storage/utils"
 	"github.com/vulcanize/vulcanizedb/libraries/shared/transformer"
 	"github.com/vulcanize/vulcanizedb/pkg/datastore/postgres"
-	"time"
 )
 
 type IStorageWatcher interface {
 	AddTransformers(initializers []transformer.StorageTransformerInitializer)
-	Execute(diffsChan chan utils.StorageDiff, errsChan chan error, queueRecheckInterval time.Duration)
+	Execute(queueRecheckInterval time.Duration, backFillOn bool)
+	BackFill(backFiller storage.BackFiller)
 }
 
 type StorageWatcher struct {
 	db                        *postgres.DB
 	StorageFetcher            fetcher.IStorageFetcher
 	Queue                     storage.IStorageQueue
 	KeccakAddressTransformers map[common.Hash]transformer.StorageTransformer // keccak hash of an address => transformer
+	DiffsChan                 chan utils.StorageDiff
+	ErrsChan                  chan error
+	BackFillDoneChan          chan bool
+	StartingSyncBlockChan     chan int
 }
 
-func NewStorageWatcher(fetcher fetcher.IStorageFetcher, db *postgres.DB) StorageWatcher {
+func NewStorageWatcher(f fetcher.IStorageFetcher, db *postgres.DB) *StorageWatcher {
 	queue := storage.NewStorageQueue(db)
 	transformers := make(map[common.Hash]transformer.StorageTransformer)
-	return StorageWatcher{
+	return &StorageWatcher{
 		db:                        db,
-		StorageFetcher:            fetcher,
+		StorageFetcher:            f,
+		DiffsChan:                 make(chan utils.StorageDiff, fetcher.PayloadChanBufferSize),
+		ErrsChan:                  make(chan error),
+		StartingSyncBlockChan:     make(chan int),
+		BackFillDoneChan:          make(chan bool),
 		Queue:                     queue,
 		KeccakAddressTransformers: transformers,
 	}
 }
 
-func (storageWatcher StorageWatcher) AddTransformers(initializers []transformer.StorageTransformerInitializer) {
+func (storageWatcher *StorageWatcher) AddTransformers(initializers []transformer.StorageTransformerInitializer) {
 	for _, initializer := range initializers {
 		storageTransformer := initializer(storageWatcher.db)
 		storageWatcher.KeccakAddressTransformers[storageTransformer.KeccakContractAddress()] = storageTransformer
 	}
 }
 
-func (storageWatcher StorageWatcher) Execute(diffsChan chan utils.StorageDiff, errsChan chan error, queueRecheckInterval time.Duration) {
+// BackFill uses a backFiller to backfill missing storage diffs for the storageWatcher
+func (storageWatcher *StorageWatcher) BackFill(backFiller storage.BackFiller) {
+	// this blocks until the Execute process sends us the first block number it sees
+	startingSyncBlock := <-storageWatcher.StartingSyncBlockChan
```

Comment:
Does this get updated as the backfill is going on? Wondering what happens if the sync progresses after the backfill has been configured

Reply:
This only receives from [here](https://github.com/vulcanize/vulcanizedb/blob/statediff_at/libraries/shared/watcher/storage_watcher.go#L94) once. It is sent the blockheight for the first payload the forward stream receives to define the ending block height of the backfill range.

Reply:
Can't seem to respond to that individual comment but in regards to
"So does this mean that backfilling is executing concurrently with processing new rows? Curious if the node is likely to be publishing to both channels at once" 
the answer is yes the backfill and forward sync happen concurrently much like how the header sync progresses at the head and tail. I can confirm the geth node publishes to both channels concurrently from my tests with the super node backfill which uses the same patterns, but obviously this still needs to be tested in the same capacity once arch1 is setup.